### DISCUSSION TITLE
Suggestion: re-implement OKL language parser using the PEG parser generator "cpp-peglib"

### DISCUSSION THREAD

**Discussion by pdhahn** [libocca/occa#600] [Ideas]
I have implemented languages using `cpp-peglib` ([https://github.com/yhirose/cpp-peglib](url)). Based on experience, I feel it makes doing so straightforward, even relatively easy. Utilized in OCCA, it could improve transparency and extensibility which in turn can improve prospects for further contributions and enhancements.

This is probably just a long-term suggestion to consider at the next revamp cycle/opportunity for OCCA.

The present OKL parser is fast. A likely downside of `cpp-peglib` would be a decrease in parser performance, but the extent of that would have to be evaluated, of course.

**Comment by mkbosmans** [comment]
How does cpp-peglib compare to something like ANTLR? Any thought on which would be the best fit for OCCA?

**Reply by pdhahn** [reply]
Sorry but I am not hands-on with ANTLR so I am not really qualified to say much about it. However, a number of years ago, when I was beginning my current project which includes a custom, special-purpose programming language, I did superficially review ANTLR. After that and investigating various other tools, I was attracted to PEG-based tools for their simplicity, and after some more research I chose to use `cpp-peglib`. Now years later, I am very happy I did because it is just so very easy to use -- trivial to write the grammar rules and easy enough to write corresponding code generation for each rule. The latter is because you have a number of choices on how to handle things: you can utilize AST's or direct (combinations of) _semantic values_ and/or designated "token" values. Note I happen not to use AST's because my parser directly generates a custom linear intermediate code that is later translated. Unlike ANTLR which supports powerful _semantic_ predicates (so I read), `cpp-peglib` supports _syntactic_ predicates which still makes practical disambiguation straightforward (e.g., my programming language grammar uses the _and_-predicate in appropriate rules to distinguish boolean values "true" and "false" from variable or function names that begin with "true" or "false").

Nevertheless, I may have made a mistake recommending `cpp-peglib` to you for this reason: its current version requires C++ 17. In the history I related above, I forked an earlier version that was happy with C++ 11. But there are other PEG-based tools you could consider that may be both current in terms of their functionality yet also happy with being compiled by older C++ versions. Just search on github!

Note PEG is well described by the corresponding Wikipedia [page.](https://en.wikipedia.org/wiki/Parsing_expression_grammar) In my experience, PEG's feature of evaluating _choice rules_ in an ordered way is a big reason why I am able to use `cpp-peglib` so easily in my project's programming language, with what amounts to (IMHO) a fairly non-trivial grammar.
